src/plugins/expressions/common/execution/execution.ts

```diff
@@ -202,147 +205,143 @@ export class Execution<
+                defaultMessage: `Function {fnName} is disabled.`,
+                values: {
+                  fnName,
+                },
+              }),
+            });
+          }
 
-        return error;
-      }
-    }
+          if (this.execution.params.debug) {
+            link.debug = {
+              args: {},
+              duration: 0,
+              fn: fn.name,
+              input: currentInput,
+              success: true,
+            };
+          }
 
-    return input;
+          const timeStart = this.execution.params.debug ? now() : 0;
+
+          // `resolveArgs` returns an object because the arguments themselves might
+          // actually have `then` or `subscribe` methods which would be treated as a `Promise`
+          // or an `Observable` accordingly.
+          return this.race(this.resolveArgs(fn, currentInput, fnArgs)).pipe(
+            tap((args) => this.execution.params.debug && Object.assign(link.debug, { args })),
+            switchMap((args) => this.race(this.invokeFunction(fn, currentInput, args))),
+            switchMap((output) => (getType(output) === 'error' ? throwError(output) : of(output))),
+            tap((output) => this.execution.params.debug && Object.assign(link.debug, { output })),
+            catchError((rawError) => {
+              const error = createError(rawError);
+              error.error.message = `[${fnName}] > ${error.error.message}`;
+
+              if (this.execution.params.debug) {
+                Object.assign(link.debug, { error, rawError, success: false });
+              }
+
+              return throwError(error);
+            }),
+            finalize(() => {
+              if (this.execution.params.debug) {
+                Object.assign(link.debug, { duration: now() - timeStart });
+              }
+            })
+          );
+        })
+      ) as Parameters<Observable<unknown>['pipe']>),
+      catchError((error) => of(error))
+    );
   }
 
-  async invokeFunction(
+  invokeFunction(
     fn: ExpressionFunction,
     input: unknown,
     args: Record<string, unknown>
-  ): Promise<any> {
-    const normalizedInput = this.cast(input, fn.inputTypes);
-    const output = await this.race(fn.fn(normalizedInput, args, this.context));
-
-    // Validate that the function returned the type it said it would.
-    // This isn't required, but it keeps function developers honest.
-    const returnType = getType(output);
-    const expectedType = fn.type;
-    if (expectedType && returnType !== expectedType) {
-      throw new Error(
-        `Function '${fn.name}' should return '${expectedType}',` +
-          ` actually returned '${returnType}'`
-      );
-    }
+  ): Observable<any> {
+    return of(input).pipe(
+      map((currentInput) => this.cast(currentInput, fn.inputTypes)),
+      map((normalizedInput) => fn.fn(normalizedInput, args, this.context)),
+      // switchMap((normalizedInput) => this.race(of(fn.fn(normalizedInput, args, this.context)))),
```

Comment:
What happened to the `race`? Is it not needed?

Reply:
Thanks for the notice! I commented that while debugging and missed that place before committing. It's been fixed now.